### DISCUSSION TITLE
Move from rgba8 internally to rgba32f

### DISCUSSION THREAD

**Discussion by Quackdoc** [woelper/oculante#154] [Ideas]
Currently oculante uses `rgba8` internally. this is fine for a general purpose image viewer, but oculante has a lot of features that make it suitable for use in professional environments. however for a lot of those environments the internal 8bit handling makes this unusable. migrating to `rgba32f` will open the doors for a lot of workloads, Currently you can already use oculante to for getting specific RGB values. these values are always limited to `0-255`. 

one such workload is when verifying HDR mapping it's useful to be able to view the images RGB values, however for HDR 10bit+ is needed, and often working with 16bit images (thanks to `uint16`) which have the ranges of `0-1023` and `0-65535` respectively. and the rounding error can be quite significant. It is useful for many users to be working with 0-255. so it would be optimal to have a way to balance both of course.

working with high bitdepths internally opens the doors for a lot of use cases, from being necessary to HDR support to image verification.

**Comment by woelper** [comment]
Thanks! This is really high up on my wishlist too. Here are a couple of thoughts:

- The editing operations are already in float and are only reduced to 8 bit in the end. There should be almost no code change necessary in that area.
- rgba32 needs much more texture memory and system memory. Not sure if this is really problematic for the average user, but it could be significant for very large images.
- Relating to above, I've thought about adding two modes, 8 and 32 bit. This would of course make matters more complicated.
- This would potentially gamma issues like https://github.com/woelper/oculante/issues/91

**Reply by Quackdoc** [reply]
I could certainly see 32 being an issue on some very low end users, or for very large images like the one I posted with the black loading issue. 

at the very least, I do have a machine or two that can be used for preformance testing, an old N3050 laptop, it's pretty low end all things considered, it's no RPI but it should give a decent test once the black screen issue is sorted.

**Reply by woelper** [reply]
I started this on a branch, but I was not super satisfied as I was searching for a way to keep 8 bit as it is and only use higher bit depth where actually needed. As a next step, I will refactor every RgbaImage to a GenericImage, which should make the float transition easier in the future.

**Reply by BitCourier** [reply]
From a performance standpoint, even using rgba8 instead of the native image file format seems to be a performance degradation. This hits especially hard when dealing with huge grayscale images (as I do). 

What about using "polymorphism" to get the best of all worlds? If the loaded image is 8 Bit, there is no need to use float32 over uint8, or do I get it wrong? If the image is grayscale, only convert to rgb(a) if any operation used needs it. Of course there is a big downside: code complexity.

At least this would be the way I would go.

**Reply by woelper** [reply]
I've indeed started to use a GenericImage instead of Rgba8 on a branch. Exactly as you say, this comes with code complexity as you have to match/test the image for it's contents everywhere you use it, and implement different operators for it or use different hardware texture formats to make use of the speed gains or memory improvements. 

My strategy would be to make the container generic at first and convert to Rgba8 earlt, then gradually change code to accept the generic image and use optimisations.

**Reply by B-LechCode** [reply]
@woelper would you mind if I would help you with that? I've got grayscale images which blow my System RAM away (40 GB memory usage). I could really benefit from this :)

**Reply by woelper** [reply]
Wow, this is almost a year old now :) I still think this should be done, and I would love some help!

I've stated a new `dynamicimage` branch and got as far as getting it to compile with every image sent as `DynamicImage`. 

I've updated the `ImageExt` trait in utils.rs to work for `DynamicImage`, but we would need to use actual texture settings to benefit from it. For example if we receive a greyscale image, we should create a greyscale image with the appropriate image format. Maybe this is something you could help with? If you don't have already, I can give you access to work on the branch directly.

I would try to make the current_image in the state also a `DynamicImage` and make sure image processing works with it. For that, I would probably convert the image into rgba float and back too keep it flexible and simple.

**Reply by B-LechCode** [reply]
Most of the image processing should be easily adaptable for varying channel sizes and bit depths. Converting to 4 channels and back would still pose memory shortage problems. But as a starter it's a perfect idea!

I will have a look at it and will come back to you when I have time :)

**Reply by woelper** [reply]
On some of the pixel operations this might be easier, but at least some libraries require rgba images. Even for pixel ops, there are a couple of optimisations, for example processing them in chunks of four for rgba to make use of SIMD extensions. Making them generic for every type of image buffer could end up being quite difficult.

**Reply by B-LechCode** [reply]
We could filter the operations depending on the type. If you would want operations requiring rgba images, you would need to place an "Convert to RGBA" operation before.

**Reply by woelper** [reply]
That is a good idea. It might also be practical to auto-convert rgb8 to rgba8 for editing. Many users might find it confusing if they load a jpeg and can't edit it. Some other takeaways from my progress in the branch so far:

- while we save memory on rgb8 or gray, we increase it for HDR images. The good stuff is that we gain more flexibility in the end.
- now, HDR images are tone mapped and converted to rgba8. Once we leave them as HDR, tone mapping should happen for the texture, maybe as a shader. If we don't do that, some images are not visible at all.
- the image channel selector should now show the actual channels present in the image
- now I pass the underlying mutable rgba buffer (if any) to the editing functions. Later, as you say, we need to branch if we want to reap the benefits or insert a to rgba8 operator.
